File: packages/python/src/teleglance/parsing/registry.py
"""Pluggable media-block parsers.

A block parser is a callable that receives the whole message node and returns
the media objects it recognizes (or an empty list). The registry runs every
parser over every message; a parser that raises is logged and skipped rather
than failing the message, so markup drift in one block type never takes down
the rest.

Extending without forking::

    from teleglance.parsing import default_registry

    registry = default_registry()

    def parse_gift(node):  # hypothetical new t.me block
        block = node.css_first(".tgme_widget_message_gift")
        return [Unsupported(raw_html=block.html)] if block else []

    registry.register("gift", parse_gift)
    client = TeleGlanceClient(registry=registry)
"""

from __future__ import annotations

import logging
from collections.abc import Callable

from selectolax.lexbor import LexborNode

from ..models.media import Media

logger = logging.getLogger("teleglance")

BlockParser = Callable[[LexborNode], list[Media]]


class ParserRegistry:
    def __init__(self) -> None:
        self._parsers: dict[str, BlockParser] = {}

    def register(self, name: str, parser: BlockParser, *, override: bool = False) -> None:
        if name in self._parsers and not override:
            raise ValueError(f"parser {name!r} already registered (pass override=True)")
        self._parsers[name] = parser

    def unregister(self, name: str) -> None:
        self._parsers.pop(name, None)

    def names(self) -> list[str]:
        return list(self._parsers)

    def copy(self) -> ParserRegistry:
        clone = ParserRegistry()
        clone._parsers = dict(self._parsers)
        return clone

    def extract(self, message_node: LexborNode) -> list[Media]:
        media: list[Media] = []
        for name, parser in self._parsers.items():
            try:
                media.extend(parser(message_node) or [])
            except Exception:
                logger.warning("block parser %r failed on a message", name, exc_info=True)
        return media

File: packages/ts/src/parsing/registry.ts
/**
 * Pluggable media-block parsers.
 *
 * A block parser is a function that receives the whole message node and returns
 * the media objects it recognizes (or an empty list). The registry runs every
 * parser over every message; a parser that throws is logged and skipped rather
 * than failing the message, so markup drift in one block type never takes down
 * the rest.
 *
 * Extending without forking:
 * ```typescript
 * import { defaultRegistry } from 'teleglance';
 *
 * const registry = defaultRegistry();
 *
 * function parseGift(node: HTMLElement): Media[] {
 *   const block = node.querySelector('.tgme_widget_message_gift');
 *   return block ? [{ type: 'unsupported', rawHtml: block.outerHTML }] : [];
 * }
 *
 * registry.register('gift', parseGift);
 * const client = new TeleGlanceClient({ registry });
 * ```
 */

import type { Media } from '../models/media.js';
import type { HTMLElement } from 'node-html-parser';

export type BlockParser = (messageNode: HTMLElement) => Media[];

export class ParserRegistry {
  private parsers = new Map<string, BlockParser>();

  register(name: string, parser: BlockParser, opts?: { override?: boolean }): void {
    if (this.parsers.has(name) && !opts?.override) {
      throw new Error(
        `Parser '${name}' already registered (pass override: true to replace)`,
      );
    }
    this.parsers.set(name, parser);
  }

  unregister(name: string): void {
    this.parsers.delete(name);
  }

  names(): string[] {
    return Array.from(this.parsers.keys());
  }

  copy(): ParserRegistry {
    const clone = new ParserRegistry();
    clone.parsers = new Map(this.parsers);
    return clone;
  }

  extract(messageNode: HTMLElement): Media[] {
    const media: Media[] = [];
    for (const [name, parser] of this.parsers) {
      try {
        const result = parser(messageNode);
        if (result) {
          media.push(...result);
        }
      } catch (err) {
        console.warn(`Block parser '${name}' failed on a message:`, err);
      }
    }
    return media;
  }
}
